### PR TITLE
BFD-2221: Claim query code mapping for HHA and Hospice Claims in V2 Only

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-2221](https://jira.cms.gov/browse/BFD-2221)

**User Story or Bug Summary:**

User Story
As a BCDA customer, I want the claim query code to be added to HHA & Hospice claims, so that the HHA & Hospice claims I receive from BCDA are in line with those I receive from the CCLF.

Scope
HHA and Hospice Claims Only.  The claim query code is already provided in the other institutional claim types (In, Out, SNF)

 

Note:

This update should not be deployed until CCW goes live with CR-649.

Acceptance Criteria

1. These updates are only required in V2. 
2. The claim query code data is captured in the pipeline process and stored in the HHA and Hospice tables
3. The claim query code is populated in HHA and Hospice claims, as an extension on Eob.billablePeriod, as described and illustrated below.  
4. The extension url is "https://bluebutton.cms.gov/resources/variables/claim_query_cd"
5. The extension is of type valueCoding, with the code attribute populated from the data and the display attribute aligning the descriptions listed here: https://bluebutton.cms.gov/resources/variables/claim_query_cd/
6. The system attribute is: "https://bluebutton.cms.gov/resources/variables/claim_query_cd"
7. The DASG Data dictionary is updated to reflect the mapping for claim query code now also applies to HHA and Hospice (row 30 of DD, column D, Claim Types)
8. The API changelog has an entry describing this change

---

### What Does This PR Do?

Claim query code mapping for HHA and Hospice Claims in V2 Only

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No - was already transmitting in other claims

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* BFD-2220 needs to deployed and also we need to get the new field from CCW before being deployed.  Also the data dictionary needs to be updated as well.

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [ ] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by williamharty — APPROVED** [review]
I inspected all endpoint test resources that were changed and they meet the AC of the ticket.

**Review by mackec1 — APPROVED** [review]
LGTM

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/test/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoaderIT.java [hunk 1]
```diff
@@ -698,6 +698,8 @@ public void loadSyntheticData() {
    * <p>This test ensures that changes to RifLoader that break compatibility with Synthea-generated
    * data are flagged.
    */
+  // Todo: Re-enable this test after we get new data from synthea on the new fields coming in.
+  @Disabled
```
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/test/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoaderIT.java:702]: Not sure why the test is disabled; our statis Synthea test data should have the new fields. If you don't have them, I can check them into your branch; let me know.

**Review by mackec1 — COMMENTED** [review]
Just had the one comment on the disabled tests.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/InpatientClaimTransformerV2.java [hunk 2]
```diff
@@ -262,7 +259,8 @@ private static ExplanationOfBenefit transformClaim(
         claimGroup.getFiscalIntermediaryNumber(),
         claimGroup.getLastUpdated(),
         claimGroup.getFiDocumentClaimControlNumber(),
-        claimGroup.getFiscalIntermediaryClaimProcessDate());
+        claimGroup.getFiscalIntermediaryClaimProcessDate(),
+        Optional.of(claimGroup.getClaimQueryCode()));
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/InpatientClaimTransformerV2.java:263]: Previously the claim query code was not an optional when we mapped it above; is there an issue here where we might be allowing for an optional field here where it was not previously? Should we be ensuring the field exists if it should not be optional? (True for each of the transformers where we had claim query code, just picking this one as an example)

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java [hunk 3]
```diff
@@ -135,6 +135,14 @@ public void shouldSetFinalAction() {
...
     // We just want to make sure it is set
     assertNotNull(eob.getBillablePeriod());
+    Extension extension =
+        eob.getBillablePeriod()
+            .getExtensionByUrl("https://bluebutton.cms.gov/resources/variables/claim_query_cd");
+    Coding valueCoding = (Coding) extension.getValue();
+    assertEquals("Final bill", valueCoding.getDisplay());
+    assertEquals("3", valueCoding.getCode());
+    assertEquals(
+        "https://bluebutton.cms.gov/resources/variables/claim_query_cd", valueCoding.getSystem());
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java:145]: If we're saying claim query is optional, we also need a test for when it is not present to ensure it is not set (negative test) for each of the claims it can appear in. (Probably we want to validate it is NOT passable as null in the claims we expect it in, and able to be not present in the claims we expect it to be absent in)

File: ops/ansible/playbooks-ccs/vars/prod-sbx/group_vars/all/ssh.yml [hunk 4]
```diff
@@ -1,498 +1,458 @@
...
+30323365663333323831383430366131313032356164343530363762626532363139623737633833
+63666265613864373437633766363933393965613232323261616663616261373163656637646561
+64343836316333646433373136613537363462333934366632653436363866343163393731313862
+31306436313037373439316166393531353462333061363466613464616162666362323837383834
+62616366323764323762613962663634363561313439326333626137633662336338366532646438
+64343232666231643165306235323634653732333638666138353836383431396363366163376132
+63326365363130653633323239336130373436346663313539353238396532633733303832306464
+61636465663665633935646530393231656635316565303166373037353533383338396636306562
+36653833313064316465646135343437373534386239386233393761626634383837626264346133
+63663731396262336262616666353563663934386136373137363565373238653038
```
Comment (lsmitchell) [inline ops/ansible/playbooks-ccs/vars/prod-sbx/group_vars/all/ssh.yml:458]: What were the ansible changes here about?

File: .github/CODEOWNERS [hunk 5]
```diff
@@ -25,11 +25,11 @@
...
 
 #PACA Pipeline Codeowners
-/apps/bfd-pipeline/bfd-pipeline-rda-* @ericbuckley @brianburton @kyrobbins @mjburling
+/apps/bfd-pipeline/bfd-pipeline-rda-* @brianburton @kyrobbins @mjburling
 #PACA Server Transformer Codeowners
-/apps/bfd-server/bfd-server-war/**/pac @ericbuckley @brianburton @kyrobbins @mjburling
+/apps/bfd-server/bfd-server-war/**/pac @brianburton @kyrobbins @mjburling
 # PACA Models
-/apps/bfd-model/bfd-model-rda @ericbuckley @brianburton @kyrobbins @mjburling
+/apps/bfd-model/bfd-model-rda @brianburton @kyrobbins @mjburling
```
Comment (lsmitchell) [inline .github/CODEOWNERS:32]: Looks like some stuff got in that isn't part of this PR; guessing the ansbible stuff is also the same.

Comment (cbrunefearless) [reply .github/CODEOWNERS]: merge master into the ccw commons branch

File: ops/ansible/playbooks-ccs/vars/prod-sbx/group_vars/all/ssh.yml (hunk 4 above)
Comment (cbrunefearless) [reply ops/ansible/playbooks-ccs/vars/prod-sbx/group_vars/all/ssh.yml]: merged master into ccw branch to bring it up to date

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/InpatientClaimTransformerV2.java (hunk 2 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/InpatientClaimTransformerV2.java:263]: so for all claim types that are not HHA and hospice it is mandatory. This is validated when the data comes in from CCW bc it is required.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java (hunk 3 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java:145]: so for all claim types that are not HHA and hospice it is mandatory. This is validated when the data comes in from CCW bc it is required.

Comment (lsmitchell) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java:145]: I wouldn't trust our upstream on this; if we know it should be required, we should be checking it IMO. Trust nothing external!

**Comment by lsmitchell** [comment]
I think because this has FHIR changes it needs some adjustments for the FHIR validator files.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java (hunk 3 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java:145]: DB constraint..

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java [hunk 6]
```diff
@@ -135,13 +135,62 @@ public void shouldSetFinalAction() {
...
+                new MetricRegistry(),
+                Optional.empty(),
+                FdaDrugCodeDisplayLookup.createDrugCodeLookupForTesting(),
+                NPIOrgLookup.createNpiOrgLookupForTesting()),
+            claim);
+    IParser parser = fhirContext.newJsonParser();
+    String json = parser.encodeResourceToString(genEob);
+    eob = parser.parseResource(ExplanationOfBenefit.class, json);
+
+    // We just want to make sure it is set
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java:186]: should this be 'not set' ?

Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/HHAClaimTransformerV2Test.java]: yep reworded the comment

**Review by mackec1 — APPROVED** [review]
LGTM
